### ISSUE TITLE
Reads done on behalf materialized view-updates have unlimited concurrency

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#15805]
Reads done as part of generating view updates -- both the reading of the base update itself, as well as the read of the pre-update state (if needed) -- are done with unlimited concurrency. We only create tracking permits for these and on numerous occasions, these were observed to accumulate in memory by the 10s of thousands, consuming hundreds of MiBs of memory.

The solution is to make these readers obtain the permit via the regular, admission path. That way, their concurrency as well as their memory consumption would have an upper bound.
Extra care must be taken however to study the impact of this change on the regular read path. We must make sure views do not flood the read semaphore to such a degree, that regular reads cannot make progress.

**Comment by denesb** [comment]
@nyh can you please link any already opened issues that might be related?

**Comment by mykaul** [comment]
https://github.com/scylladb/scylladb/issues/15717 ?

**Comment by nyh** [comment]
This is a duplicate of https://github.com/scylladb/scylladb/issues/8873, although the exact symptoms of the problem changed over the years: Initially, view reads got limited together with all reads, and caused all reads to slow down. Today view reads aren't limited, so they can cause an OOM.

We also had other issues over the years that may be related or partially related (some are already closed as duplicates, there are so many in this list I didn't consider each one very carefully):
* https://github.com/scylladb/scylladb/issues/7848
* https://github.com/scylladb/scylladb/issues/8919
* https://github.com/scylladb/scylladb/issues/11961
* https://github.com/scylladb/scylladb/issues/11963
* https://github.com/scylladb/scylladb/issues/12379
* https://github.com/scylladb/scylladb/issues/14909
* https://github.com/scylladb/scylladb/issues/15717
* https://github.com/scylladb/scylla-enterprise/issues/1280
* https://github.com/scylladb/scylla-enterprise/issues/1783

**Comment by mykaul** [comment]
@denesb - this needs to get priority - QA can't really test MVs without hitting it, I believe customers are suffering from it as well.

**Comment by denesb** [comment]
There are two options here:
* I do it. I am very familiar with the problem and the infrastructure, but I have very little time to code these days, with constant interruptions. I  would probably take months to deliver this.
* Someone else (who is not already on the tablets project) does it, with supervision and help from me. Preferably someone who is at least familiar with MV code. They would need some ramp-up time and there would probably be many iterations of the resulting PR, but there is a good chance that this is still the faster option.

**Comment by mykaul** [comment]
> There are two options here:
> 
> * I do it. I am very familiar with the problem and the infrastructure, but I have very little time to code these days, with constant interruptions. I  would probably take months to deliver this.
> * Someone else (who is not already on the tablets project) does it, with supervision and help from me. Preferably someone who is at least familiar with MV code. They would need some ramp-up time and there would probably be many iterations of the resulting PR, but there is a good chance that this is still the faster option.

I prefer the latter. @bhalevy ?

**Comment by bhalevy** [comment]
I've no objection. But why can't the MV team own this.
If the interface to the concurrency semaphore isn't clear enough to be used by somebody other than its creator, I prefer Botond to improve its documentation rather implementing stuff using it while he's the only one familiar with it.

**Comment by mykaul** [comment]
> I've no objection. But why can't the MV team own this. If the interface to the concurrency semaphore isn't clear enough to be used by somebody other than its creator, I prefer Botond to improve its documentation rather implementing stuff using it while he's the only one familiar with it.

@kostja ?

**Comment by roydahan** [comment]
Hit again during create_index nemesis in run https://argus.scylladb.com/test/325d1aa5-49f7-4a12-920b-46406d16513f/runs?additionalRuns[]=3404ba45-6832-4891-8ba9-5f61757d5c9c.

**Comment by eliransin** [comment]
I am going to do it.
I will sync with @denesb  on this ASAP.
After reviewing the code, it looks like it is a simple matter of making permits instead of permits for tracking.
The only problem with this is that it will "tie" the write path and read paths together (at least for base table writes).
We acquire a lock on some row before reading it. If we then wait on a semaphore, we are going to be blocking this row for possibly `timeout` time.
So we will probably need to revert the order (enter semaphore, then acquire lock).
I will investigate some more and come up with an implementation (and more importantly) testing plan for this.

One thing that bothers me is this, it seams that we must compromise on something:
1. If those reads are limited, they are going to limit writes on those tables, I think it is inevitable but it might present a change in behavior.
2. One thing that should be taken into account is, if we don't create a separate path for those reads (different semaphore), we are going to end up with view reads blocking normal user reads, however I don't think it is bad, our system can handle a certain amount of reads and writes and we need to share them between users.

My intuition for this is that if we would charge users by op, a base table write would probably cost write*num_of_views + read.

**Comment by nyh** [comment]
> I am going to do it. I will sync with @denesb on this ASAP. After reviewing the code, it looks like it is a simple matter of making permits instead of permits for tracking.

Please also read the discussion and work by myself, Piotr Sarna and Jan, in https://github.com/scylladb/scylladb/issues/8873. I also believe that the fix will be simple, BUT, in my mind any fix will be worthless without a good test to reproduce the problem and showing the fix fixes it. As you can see Piotr Sarna worked on such a reproducer, but we never had the opportunity to finish that work.

> The only problem with this is that it will "tie" the write path and read paths together (at least for base table writes).

This is not a problem because writes to base tables (which have views) are **already** tied to read, namely the "read before write". There's even a row lock already. It's not a new thing.

What we do need to be very careful not to do (and this is what #8873 was originally about) is to slow down regular user reads just because there are a lot of ongoing writes. We don't want 3 user reads plus 1 million MV-read-before-writes to wait on the same semaphore ;-)

> One thing that bothers me is this, it seams that we must compromise on something:
> 
>     1. If those reads are limited, they are going to limit writes on those tables, I think it is inevitable but it might present a change in behavior.

I don't think this is a problem. The reason why we can have a million of those base-replica writes (note this is NOT the user's coordinator-side writes!) is because this is what our regular (not MV!) "throttling" does. They are already a huge queue of writes that will happen eventually and slowly - it's fine for them to wait even longer.

>     2. One thing that should be taken into account is, if we don't create a separate path for those reads (different semaphore), we are going to end up with view reads blocking normal user reads, however I don't think it is bad, our system can handle a certain amount of reads and writes and we need to share them between users.

This would be very bad, and in fact this was the original reason why #8873. In the old days we didn't have OOMs like we had today - rather we had terrible read performance for user read requests, and we don't want to repeat this.

Please remember (and read #8873 for the full discussion) that even if the client has write concurrency of 10, we can end up - because of the "throttling" algorithm we use (unrelated to MV) - with one of the replicas doing a million writes in parallel, and therefore a million reads in parallel. It is not true that we'll only have this problem if the client has absurd concurrency in its own writes (if the client does have high write concurrency, it's indeed a problem we can't solve, and don't need to solve).

**Comment by denesb** [comment]
I think we will indeed have to introduce a new semaphore for these reads, otherwise read-before-writes done on behalf of MVs will possibly starve user reads. This new semaphore can have a much more limited concurrency and memory pool -- it is fine to make MV updates wait a bit more.

**Comment by avikivity** [comment]
Yes it needs a separate resource pool. It can't be too small, or write throughput will suffer.

**Comment by eliransin** [comment]
The information in https://github.com/scylladb/scylladb/issues/8873 is quite extensive and I think I have a good grasp of what is going on, expecially after discussing with @nyh about it.

We have an (imaginary) write buffer that is only limited by memory (there is no write_concurrency semaphore) on the replica side, writes to base tables bind objects in this buffer to read operations which makes them a lot slower than other writes in the buffer which in turn suggests that in steady state, all of the buffer will be filled with base table writes. Over the years several solution have been attempted, the current "solution" is to do nothing.
This solution is bad because it can lead to an OOO since the reads are not accounted for.
The past discussed solutions:
1. Simply create a separate concurrency reader for those, implications:
The reads (which are bound to writes) will makes those writes even slower and the buffer will still be filled with those writes in steady state. This in turn will not allow any other write to happen (in parallel) on this replica node. If those writes are roughly interleaved, another writes to a non-base table, will probably start to see latencies roughtly equivalent to the slow writes latencies (which are actually read latencies).
On the read side, whatever concurrency we choose, if it saturates the disk, we are expected to see the user reads roughly doubled, if the user concurrency is very low (less than what we choose for the semaphore) we might see latencies increase by a factor, in general my congesture for the impact is that the this factor is `min(user_concurency,read_semaphore_concurrency) / [ min(user_concurency,read_semaphore_concurrency) + views_semaphore_concurrency ]` , just as an example, if the user has a concurrency of 1 and the views_semaphore has 100 cocurrency units the user is expected to experience a latency of `normal_read_latency * 1 / 101` during this state.

2. Use the "normal" reader concurrency semaphore - this is even worse it will also make user reads suffer.



All of the above still needs some proving but there are evidence for that accross the aforementioned issue, for example,
sarna found out empirically that he has to set the concurrency for the view to a very low number (4 in his case) in order to reap the isolation benefits of the semaphore which is probably a consequesce of this factor.

For conclusion, I believe that the separaion sould be done on more levels like io scheduling for example (I think this is maybe what @avikivity meant in https://github.com/scylladb/scylladb/issues/15805#issuecomment-1875673896 when he said )

My concrete plan is:
1. Tailor a case where we see the current problem (where we do nothing to limit base table read before writes)
2. Implement a semaphore for the view reads (naively) by using the "user reads" semaphore parameters.
3. Try to prove the conjesture above by expanding the scenario from 1
4. If I can't prove it, this will be considered as an extra testing and I will submit the PR, If I can
5. Consider the next step which will probably include separating base table writes and normal writes into separate scheduling groups (so IO scheduling will not be an issue).

**Comment by denesb** [comment]
@eliransin I fully agree that this effort should be lead by experiments. It is very hard to predict how a certain solution will behave so we have no way around doing reproducer-driven-development here.

**Comment by eliransin** [comment]
Update: 
I have set up a basic local setup which include 3 nodes, each with smp=1 and memory=2G, I ran cassandra-stress several times and ran once into the "bad-alloc" exception.
However, just running the cluster with base-table writes on two nodes serving as coordinators is too flaky.
I did witness several millions of permits accumulating on the nodes, presumably those are the MV tracking only read before writes.
I am now going  to try and design an experiment using `cgroups` where I will run the cluster at a steady state and then,
start to throttle one of the nodes cpu and observe how it affects his background writes accumulation.
Hopefully, I will come up with a stable enough scenario that can be reproduced fast and reliably and then I will start to test
the seemingly obvious solution of read concurrency semaphore.

**Comment by eliransin** [comment]
New plan, after discussing with @dorlaor and @mykaul about this:
I am going to add the semaphore with equal memory and concurrency to the user writes semaphore, in addition,
going to add two new configuration options:
1. To enable the use of this semaphore (in non tracking only mode) - live updatable and on by default
2. To change the concurrency parameter of the semaphore - requiers restart (default is the same as the user's read concurrency semaphore)

Reason:
It has been proven hard to reproduce reliably and even worse, we have no idea how to explore the ripples of this changeon other parts of the system.
We are quite convinced that admission control and separation  is at least part of the solution, we also know that we 
have more than one SCT that reproduces the issue. This will be able to give us a quick feedback about this while not
compromising having a regression down the road with real users use case. After it is battle tested and we are convinced that there are not unexpected consequences from this change we can remove the configuration parameters.

**Comment by mykaul** [comment]
@piodul - is this on track for 6.0, or shall we move it to 6.1?
